src/terminal/terminal.go

```diff
@@ -0,0 +1,192 @@
+// toast   be->fe     Data           OOB message to be shown to the user
+type TerminalMessage struct {
+	Op, Data, SessionID string
+	Rows, Cols          uint16
+}
+
+// startProcess executes cmd in the container and connects it up with the ptyHandler (a session).
+func startProcess(k8sClient kubernetes.Interface, cfg *rest.Config, clusterName, namespaceName, podName, containerName string, cmd []string, ptyHandler PtyHandler) error {
+	req := k8sClient.CoreV1().RESTClient().Post().
+		Resource(resource).
+		Name(podName).
+		Namespace(namespaceName).
+		SubResource(operation)
+	req.VersionedParams(&v1.PodExecOptions{
+		Container: containerName,
+		Command:   cmd,
+		Stdin:     true,
+		Stdout:    true,
+		Stderr:    true,
+		TTY:       true,
+	}, scheme.ParameterCodec)
+
+	exec, err := remotecommand.NewSPDYExecutor(cfg, http.MethodPost, req.URL())
+	if err != nil {
+		return err
+	}
+
+	ctx := multicluster.WithMultiClusterContext(context.Background(), clusterName)
+	err = exec.StreamWithContext(ctx, remotecommand.StreamOptions{
+		Stdin:             ptyHandler,
+		Stdout:            ptyHandler,
+		Stderr:            ptyHandler,
+		TerminalSizeQueue: ptyHandler,
+		Tty:               true,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GenTerminalSessionId generates a random session ID string.
+// This ID is used to identify the session when the client opens the WebSocket connection.
+func GenTerminalSessionId() (string, error) {
+	bytes := make([]byte, 16)
+	if _, err := rand.Read(bytes); err != nil {
+		return "", err
+	}
+	id := make([]byte, hex.EncodedLen(len(bytes)))
+	hex.Encode(id, bytes)
+	return string(id), nil
+}
+
+// isValidShell checks if the shell is an allowed one
+func isValidShell(validShells []string, shell string) bool {
+	for _, validShell := range validShells {
+		if validShell == shell {
+			return true
+		}
+	}
+	return false
+}
+
+// WaitForTerminal is called from capp controller as a goroutine.
+// Waits for the WebSocket connection to be opened by the client the session to be Bound in handleTerminalSession.
+// After the session is bound, it manages the session by opening the stream and closing it in case of an error or timeout.
+func WaitForTerminal(k8sClient kubernetes.Interface, cfg *rest.Config, clusterName, namespaceName, podName, containerName, shell, sessionId string, logger *zap.Logger) {
+	select {
+	case <-TerminalSessions.Get(sessionId).Bound:
+		close(TerminalSessions.Get(sessionId).Bound)
+
+		var err error
+		validShells := []string{"bash", "sh", "powershell", "cmd"}
+
+		if isValidShell(validShells, shell) {
+			cmd := []string{shell}
+			err = startProcess(k8sClient, cfg, clusterName, namespaceName, podName, containerName, cmd, TerminalSessions.Get(sessionId))
+		} else {
+			// No shell given or it was not valid: try some shells until one succeeds or all fail
+			// FIXME: if the first shell fails then the first keyboard event is lost
```

Comment:
In what cases will the shell fail?
Will it fail for pods with no distro (like our controllers)?